20231214-listUsers-api.md

```diff
@@ -0,0 +1,261 @@
+  - https://github.com/openfga/openfga/issues/406
+  - https://github.com/openfga/openfga/issues/1215
+
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+The `ListUsers` and `StreamedListUsers` will provide an API that answers the question
+
+> Who are all the users that have a relationship with an object?
+
+More specifically, given an [object](https://openfga.dev/docs/concepts#what-is-an-object) and [relation](https://openfga.dev/docs/concepts#what-is-a-relation), these APIs will return all of the [user](https://openfga.dev/docs/concepts#what-is-a-user) objects of a particular type that have that [relationship](https://openfga.dev/docs/concepts#what-is-a-relationship).
+
+
+# Definitions
+[definitions]: #definitions
+
+* You may see references to strings formatted as **`object#relation@user`** in this document. This is short-hand notation for representing an OpenFGA Relationship Tuple. For example, `group:eng#member@user:jon` or `document:1#viewer@group:eng#member` etc..
+
+* **Expansion** - refers to the process of iteratively traversing the graph of relationships that are realized through both an OpenFGA model AND the relationship tuples that exist in the store.
+
+* **Forward expansion** - refers to expansion of the OpenFGA relationship graph in a forward form. That is, starting at an object and relation, walk the graph of relationships in a directed way towards a user of some specific type by following paths that would lead through some target relation.
+
+* **Reverse expansion** - refers to expansion of the OpenFGA relationship grpah in a backwards form. That is, starting at a user of a specific type, walk the graph of relationships in a directed way towards an object of some specific type by following paths that would lead through some target relation.
+
+* **Concrete/terminal objects** - a concrete or terminal object refers to a singular or specific resource of a given object type and id. A concrete object cannot be expanded any further (see Expansion above). It can be conceptualized as a leaf-node in a graph. Concrete objects differ from usersets, because usersets refer to a collection of zero or more concrete objects. For example `user:jon` is a concrete object while `group:eng#member` is a userset which may expand to multiple concrete objects such as `user:jon`, `user:andres`, etc..
+
+* **Userset** - A userset refers to a set of zero or more concrete objects that are the subjects that a relationship expands to. See our Concepts page on [Users](https://openfga.dev/docs/concepts#what-is-a-user) for more info.
+
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+
+Developers using OpenFGA want to be able to do a reverse query and lookup all of the users that have a particular relationship with an object. 
+
+- What use cases does it support?
+
+**UIs** - display the users that a resource has been shared with. Think of the "Share" dialog in Google Docs, for example.
+
+**todo: fill out more**
```

Comment:
@aaguiarz could you help me add to this list? If you want to respond to this comment, then I'll add that content to the RFC here.

Reply:
I want this endpoint because I have a project where I need to notify all users with access to an object if any changes occur to it, with said access possibly being indirect (I'm currently polyfilling it with `Expand`).

Reply:
Added.